controller/src/lib.rs

```diff
@@ -1602,30 +1602,65 @@ impl IoLoop {
                     );
                     probes::message__received!(|| (peer.ip(), &message));
 
-                    // Sanity check the protocol version.
-                    if message.header.version != message::version::CURRENT {
-                        // If the version does not match, we're choosing to drop
-                        // the packet rather than reply with a version mismatch
-                        // error. Without a matching version, we can't really
-                        // trust the message kind we have deserialized, so won't
-                        // be able to reliably send protocol errors.
+                    // Check the version in the response.
+                    //
+                    // Beginning at version `MIN` (V5), we committed to
+                    // backwards compatibility. In particular, any existing
+                    // messages must still be decodable by software running a
+                    // later version of the protocol.
+                    //
+                    // In this case we're processing a response to our own
+                    // request. If we _get_ a response at all, it means that the
+                    // peer is running at least `MIN`. There are actually no
+                    // other checks that need to be performed, which would
+                    // prevent us from decoding the message.
+                    //
+                    // Suppose the peer is running between `MIN` and `CURRENT`.
+                    // Then we can clearly process their response, because the
+                    // version is one that we've committed to compatibility
+                    // with. This could be a version-mismatch error message,
+                    // because the _peer_ may not be able to handle the message.
+                    //
+                    // If the version is _newer_ than `CURRENT`, we can still
```

Comment:
This sentence is confusing to me – if the peer is running a version that's far ahead of `CURRENT`, why can we still process their messages?  Is it because at this point, we've already bailed out with deserialization errors if we couldn't parse it?

Reply:
I think that's right. And I also think this points out an issue. But first, let me expand on this part. Here, we've received a response to our request, which we can completely deserialize. The Sad Case is when that has a version before the compatibility commitment. We should never really get that, because any Hubris code running those versions will just drop a message whose version doesn't match its own, but I'm pretty paranoid at this point.

The point I was trying to make here is that it doesn't matter what version the peer is running, as long as its at least `MIN`. If they're way ahead of `CURRENT`, they still have to be able to process and respond to all the messages that _the host side_ can send. That's because the SP's current is a superset of ours. At least, it is if we correctly maintain backwards compatibility.

I hope that helps, but let me know how you think the comment can be improved. Some of this was written as I was solidifying the ideas, so I'm sure it can be tightened up.

---

Now to the problem this raises. If you look above, I'm handling deserialization failures in the same way as before: log and drop the message, but otherwise just continue. I'm not sure that's correct. In particular, I think we'll still _retry the message_. We need the same kind of error-injection I have in the code just below your comment, starting at L1656 in the PR diff. I.e., we need to `take()` the outstanding message, and send an error on the reply channel, so that the code making the request can bubble that up. Does that sound accurate?

Reply:
What about something like this?
```rust
                    // If the version is _newer_ than `CURRENT`, we can still
                    // process it, because we successfully deserialized it
                    // (above).  Our peer has committed to this same
                    // compatibility, so we know that successful deserialization
                    // means that we share a common understanding of the message
                    // semantics with our peer.
```

The retry issue was pre-existing before this PR, right?  If that's the case, then I agree with your assessment.

Reply:
Yes, the retry issue is separate. I've updated the code to inject a failure up to the caller. I like your suggested comment, I'll work that in. Thanks!